### ISSUE TITLE
ZIO.Schedule StackOverflowError after upgrading to 1.0.0

### ISSUE THREAD

**Issue by mindartur** [zio/zio#4063]
After upgrading to ZIO 1.0.0 following error started to appear:

```
java.lang.StackOverflowError
at zio.Schedule$.$anonfun$unfold$3(Schedule.scala:1041)
at zio.Schedule$.$anonfun$unfold$3(Schedule.scala:1041)
at zio.Schedule$.$anonfun$unfold$3(Schedule.scala:1041)
at zio.Schedule$.$anonfun$unfold$3(Schedule.scala:1041)
....
at zio.Schedule$.$anonfun$unfold$3(Schedule.scala:1041)
Exception in thread "main"
```

No more info is presented :( Inside our code we only use `Schedule.forever` and `Schedule.recurs()`

**Comment by mindartur** [comment]
Seems that increasing stack size helps :D 

Will be watching and report if see this again

**Comment by mindartur** [comment]
Still happens with increase stack, but more rarely

**Comment by mindartur** [comment]
```
def spec = suite("test Schedule")(
    testM("test Schedule forever")( {
      for {
        result <- ZIO.collectAllPar((1 to 50).map(_ => ZIO(throw new Exception("exception")))).retry(Schedule.forever).timeout(Duration(30, TimeUnit.MINUTES))
      } yield assert(result)(equalTo(None))
    })
  )
```
Some nasty test, with `-Xss1m` fails almost instantly, with `-Xss10m` fails in ~5 minutes with `StackOverflowError`. 

My guess is that this function is not tail recursive: https://github.com/zio/zio/blob/41041307885a40d86a739575b8603e065cce4a14/core/shared/src/main/scala/zio/Schedule.scala#L1040

**Comment by tuleism** [comment]
Attempt to fix at #4064

**Comment by mindartur** [comment]
seems to be fixed in latest snapshot! thanks!
